### ISSUE TITLE
Digit grouping

### ISSUE THREAD

**Issue by cup** [nim-lang/Nim#11734]
Nim seems to have no way to format a number with commas or digit grouping.
Other languages offer this, for example C:

~~~c
#define _XOPEN_SOURCE 700
#include <locale.h>
#include <stdio.h>
int main() {
   setlocale(LC_ALL, "");
   printf("%'d\n", 1000);
}
~~~

Python:

~~~py
f'{1000:,}'
~~~

JavaScript:

~~~js
(1000).toLocaleString();
~~~

PHP:

~~~php
number_format(1000);
~~~

Go:

~~~go
package main
import "golang.org/x/text/language"
import "golang.org/x/text/message"
func main() {
   message.NewPrinter(language.English).Println(1000)
}
~~~

D:

~~~d
import std.format, std.stdio;
void main() {
   auto aa = "%,d".format(1000);
   writeln(aa);
}
~~~

**Comment by narimiran** [comment]
> Nim seems to have no way to format a number with commas or digit grouping.

[strutils.insertSep](https://nim-lang.github.io/Nim/strutils.html#insertSep%2Cstring%2CChar%2Cint)

**Comment by mratsim** [comment]
Probably an option to add to [strformat formatting of numbers](https://nim-lang.org/docs/strformat.html#standard-format-specifier-for-strings-integers-and-floats)

**Comment by Araq** [comment]
Well yes, but since it's deterministic, ``.replace('.', ',')`` will always work well.

**Comment by cup** [comment]
@narimiran perfect, thanks. this works for me:

~~~nim
import strutils
const aa = $(1000)
echo insertSep(aa, ',')
~~~

**Comment by cup** [comment]
@narimiran thanks, this works for me:

~~~nim
import strutils
const aa = $(1000)
echo insertSep(aa, ',')
~~~

I will keep this open if its ok, as `insertSep` works with strings, and probably
a method should be available for numbers as well as @mratsim said. something
like this:

~~~nim
import strformat
const aa = 1000
echo fmt"{aa:,}"
~~~

**Comment by qqtop** [comment]
```
import nimcx
let a = 1234567890
let b = -10007887236.4567
echo ff2(a,2)
echo ff2(b,5)
```
gives

```

1,234,567,890.00
-10,007,887,236.45670

```

**Comment by cup** [comment]
@qqtop thanks, but looks like that doesnt work on windows

**Comment by qqtop** [comment]
nimcx is made for linux only but you are free to just  extract the relevant procs
and adapt / improve them for windows . The procs sit in cxglobal.nim

**Comment by cup** [comment]
@qqtop I posted an issue on nimcx repo - i dont prefer to discuss it further here

thank you

**Comment by krux02** [comment]
just a small remark: ``(1000.5).toLocaleString();`` returns ``'1.000,5'`` on my system. I don't know if you ever want that.

**Comment by cup** [comment]
@krux02 I would prefer to stay on topic, as any JavaScript issue is clearly off
topic. However I will say that your output is completely normal. As stated in
the function name `toLocaleString` should and does produce locale-dependant
decimal separator and thousands separator.

With US locale this is `1,234.5`, with European locale this is `1.234,5`.
Thanks.

**Comment by cup** [comment]
@narimiran I just noticed an issue:

~~~nim
import strutils
const aa = $(1234.56)
echo aa.insertSep(',')
~~~

Result:

~~~
1,234,.56
~~~

**Comment by narimiran** [comment]
It is not an issue, it works as advertised: `the algorithm works with any string s`, meaning it treats `12345`, `123.45`, `abcde` completely the same.

**Comment by cup** [comment]
ok then, it seems Nim standard library currently does not have a robust solution
then. In the meantime I found this package that seems to do the job:

~~~nim
import strfmt
const aa = 1234.56
echo "{:,}".fmt(aa)
~~~

https://github.com/kaushalmodi/strfmt

**Comment by krux02** [comment]
@cup can we mark this feauter request then as resolved?

**Comment by cup** [comment]
@krux02 um, no? It would be resolved if it was in the standard library, or if the developers are currently officially refusing to do that.

**Comment by timotheecour** [comment]
for the reverse direction, see https://github.com/nim-lang/Nim/pull/15421#issuecomment-746863623

**Comment by ringabout** [comment]
See also
https://www.python.org/dev/peps/pep-0378
https://www.python.org/dev/peps/pep-0515

**Comment by timotheecour** [comment]
https://www.python.org/dev/peps/pep-0515/ is already in nim, so i guess it's mostly https://www.python.org/dev/peps/pep-0378/ to consider

**Comment by ringabout** [comment]
yeah, pep515 also introduces `_` option for f-string. Before f-string only support `,` option.

**Comment by ringabout** [comment]
using `sprintf` make it a bit hard to extend.

needs
https://github.com/nim-lang/Nim/issues/13365
to move on.
